Feature/Add Coingecko api key support

TanguyHerbron opened the pull request:
Add an environment variable to inject an API key for Coingecko datasource.

TanguyHerbron commented:
This resolves #2827

dtslvr commented on apps/api/src/services/configuration/configuration.service.ts:
```diff
@@ -15,6 +15,7 @@ export class ConfigurationService {
       BETTER_UPTIME_API_KEY: str({ default: '' }),
       CACHE_QUOTES_TTL: num({ default: 1 }),
       CACHE_TTL: num({ default: 1 }),
+      COINGECKO_API_KEY: str({ default: '' }),
```
`COINGECKO_API_KEY` -> `COINGECKO_API_KEY_DEMO`

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -21,11 +21,14 @@ import got from 'got';
 @Injectable()
 export class CoinGeckoService implements DataProviderInterface {
+  private apiKey: string;
   private readonly URL = 'https://api.coingecko.com/api/v3';
```
`private headers: Headers = {};`

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -21,11 +21,14 @@ import got from 'got';
     private readonly configurationService: ConfigurationService
-  ) {}
+  ) {
+    this.apiKey = this.configurationService.get('COINGECKO_API_KEY');
```
```
if (this.configurationService.get('COINGECKO_API_KEY_DEMO')) {
  this.headers['x-cg-demo-api-key'] = this.configurationService.get('COINGECKO_API_KEY_DEMO');
}
```

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -49,7 +52,8 @@ export class CoinGeckoService implements DataProviderInterface {
-      const { name } = await got(`${this.URL}/coins/${aSymbol}`, {
+      const { name } = await got(`${this.URL}/coins/${aSymbol
+        }?x_cg_demo_api_key=${this.apiKey}`, {
         // @ts-ignore
```
`headers: this.headers,`

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -104,7 +108,8 @@ export class CoinGeckoService implements DataProviderInterface {
+        )}&to=${getUnixTime(to)
+        }?x_cg_demo_api_key=${this.apiKey}`,
         {
           // @ts-ignore
```
`headers: this.headers,`

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -165,7 +170,8 @@ export class CoinGeckoService implements DataProviderInterface {
+        )}&vs_currencies=${DEFAULT_CURRENCY.toLowerCase()
+        }?x_cg_demo_api_key=${this.apiKey}`,
         {
           // @ts-ignore
```
`headers: this.headers,`

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -216,7 +222,8 @@ export class CoinGeckoService implements DataProviderInterface {
-      const { coins } = await got(`${this.URL}/search?query=${query}`, {
+      const { coins } = await got(`${this.URL}/search?query=${query
+        }?x_cg_demo_api_key=${this.apiKey}`, {
         // @ts-ignore
```
`headers: this.headers,`

dtslvr reviewed (changes requested):
Thank you for your contribution, @TanguyHerbron.

It's fine to start with demo. I have added some feedback with ideas. I think it's cleaner to set a header instead making the url more complicated. You also have to handle the existence once. What do you think?

TanguyHerbron commented:
> Thank you for your contribution, @TanguyHerbron.
> 
> It's fine to start with demo. I have added some feedback with ideas. I think it's cleaner to set a header instead making the url more complicated. You also have to handle the existence once. What do you think?

Nice suggestions indeed, I've added them.
I didn't know about the `headers` mechanism with `got`, nice implementation.

I will then also add the `pro` feature, give me 5 minutes.

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -17,15 +17,25 @@ import {
 export class CoinGeckoService implements DataProviderInterface {
-  private readonly URL = 'https://api.coingecko.com/api/v3';
+  private headers: Headers = {};
+  private readonly URL: string = 'https://api.coingecko.com/api/v3';
```
Can you refactor to `private url = 'https://api.coingecko.com/api/v3';` and move up to keep the alphabetical order?

It is not `readonly` anymore and you don't need to set the type due to the initial value.

TanguyHerbron replied:
I left it to `readonly` since TypeScript allows [modifications under the constructor](https://www.tutorialsteacher.com/typescript/typescript-readonly) for such variable.

The added type declaration is to prevent the linter/transpiler from complaining.

Let me know if you would rather have it as `private url` either way.

Also, isn't it already alphabetically ordered that way ?

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -51,7 +61,8 @@ export class CoinGeckoService implements DataProviderInterface {
         // @ts-ignore
-        signal: abortController.signal
+        signal: abortController.signal,
+        headers: this.headers
```
Please move up to keep the alphabetical order (above `// @ts-ignore`)

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -107,7 +118,8 @@ export class CoinGeckoService implements DataProviderInterface {
           // @ts-ignore
-          signal: abortController.signal
+          signal: abortController.signal,
+          headers: this.headers
```
Please move up to keep the alphabetical order (above `// @ts-ignore`)

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
No header needed for `getQuotes()` / `search()`?

TanguyHerbron replied:
It does, thank you, that's an oversight from my side.

dtslvr commented on apps/api/src/services/data-provider/coingecko/coingecko.service.ts:
```diff
@@ -17,15 +17,25 @@ import {
     private readonly configurationService: ConfigurationService
-  ) {}
+  ) {
+    if (this.configurationService.get('COINGECKO_API_KEY_DEMO')) {
```
Maybe this is a bit cleaner:

```
const apiKeyDemo = this.configurationService.get('COINGECKO_API_KEY_DEMO');
const apiKeyPro = this.configurationService.get('COINGECKO_API_KEY_PRO');

if (apiKeyDemo) {
  this.headers['x-cg-demo-api-key'] = apiKeyDemo;
}

...
```

TanguyHerbron replied:
Indeed, will be added in upcoming commit.

dtslvr reviewed (changes requested):
Thank you for the changes. I have added a few more comments.

dtslvr reviewed (approved):
Thank you for this contribution, @TanguyHerbron! 🙌🏻 
Ready to merge now.
